packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -131,43 +131,42 @@ export function FocusScope(props: FocusScopeProps) {
   useRestoreFocus(scopeRef, restoreFocus, contain);
   useAutoFocus(scopeRef, autoFocus);
 
-  // this layout effect needs to run last so that focusScopeTree cleanup happens at the last moment possible
+  // This needs to be an effect so that activeScope is updated after the FocusScope tree is complete.
+  // It cannot be a useLayoutEffect because the parent of this node hasn't been attached in the tree yet.
   useEffect(() => {
-    if (scopeRef) {
-      let activeElement = document.activeElement;
-      let scope: TreeNode | null = null;
-      // In strict mode, active scope is incorrectly updated since cleanup will run even though scope hasn't unmounted.
-      // To fix this, we need to update the actual activeScope here
-      if (isElementInScope(activeElement, scopeRef.current)) {
-        // Since useLayoutEffect runs for children first, we need to traverse the focusScope tree and find the bottom most scope that
-        // contains the active element and set that as the activeScope
-        for (let node of focusScopeTree.traverse()) {
-          if (node.scopeRef && isElementInScope(activeElement, node.scopeRef.current)) {
-            scope = node;
-          }
+    let activeElement = document.activeElement;
+    let scope: TreeNode | null = null;
+
+    if (isElementInScope(activeElement, scopeRef.current)) {
+      // We need to traverse the focusScope tree and find the bottom most scope that
+      // contains the active element and set that as the activeScope.
+      for (let node of focusScopeTree.traverse()) {
+        if (node.scopeRef && isElementInScope(activeElement, node.scopeRef.current)) {
+          scope = node;
         }
+      }
 
-        if (scope === focusScopeTree.getTreeNode(scopeRef)) {
-          activeScope = scope.scopeRef;
-        }
+      if (scope === focusScopeTree.getTreeNode(scopeRef)) {
+        activeScope = scope.scopeRef;
       }
+    }
+  }, [scopeRef]);
 
-      return () => {
-        // Scope may have been re-parented.
-        let parentScope = focusScopeTree.getTreeNode(scopeRef)?.parent?.scopeRef ?? null;
+  // This layout effect cleanup is so that the tree node is removed synchronously with react before the RAF
+  // in useRestoreFocus cleanup runs.
+  useLayoutEffect(() => {
```

Comment:
So to make sure I understand the previous issue:
We previously handled the cleanup in a useEffect cleanup and that could potentially run after the RAF triggered by the useRestoreFocus useLayoutEffect cleanup?

Reply:
Yep, it was a race for which got scheduled first, the cleanup raf or the useeffect raf.